internal/rbd/mirror.go

```diff
@@ -0,0 +1,134 @@
+/*
+Copyright 2021 The Ceph-CSI Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package rbd
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"strings"
+
+	"github.com/ceph/ceph-csi/internal/util"
+
+	librbd "github.com/ceph/go-ceph/rbd"
+)
+
+func (rv *rbdVolume) enableImageMirroring(ctx context.Context, mode librbd.ImageMirrorMode) error {
+	util.DebugLog(ctx, "rbd: enable mirroring %s with mode %d", rv.String(), mode)
+	image, err := rv.open()
+	if err != nil {
+		return err
+	}
+	defer image.Close()
+
+	err = image.MirrorEnable(mode)
+	return err
+}
+
+func (rv *rbdVolume) disableImageMirroring(ctx context.Context, force bool) error {
+	util.DebugLog(ctx, "rbd: disable mirroring %s with force %v", rv.String(), force)
+	image, err := rv.open()
+	if err != nil {
+		return err
+	}
+	defer image.Close()
+
+	err = image.MirrorDisable(force)
+	return err
+}
+
+func (rv *rbdVolume) getImageMirroringInfo(ctx context.Context) (*librbd.MirrorImageInfo, error) {
+	util.DebugLog(ctx, "rbd: get mirroring info %s", rv.String())
+	image, err := rv.open()
+	if err != nil {
+		return nil, err
+	}
+	defer image.Close()
+
+	info, err := image.GetMirrorImageInfo()
+	return info, err
+}
+
+func (rv *rbdVolume) promoteImage(ctx context.Context, force bool) error {
+	util.DebugLog(ctx, "rbd: promote image %s with force %v", rv.String(), force)
+	image, err := rv.open()
+	if err != nil {
+		return err
+	}
+	defer image.Close()
+	return image.MirrorPromote(force)
+}
+
+func (rv *rbdVolume) demoteImage(ctx context.Context) error {
+	util.DebugLog(ctx, "rbd: promote image %s", rv.String())
+	image, err := rv.open()
+	if err != nil {
+		return err
+	}
+	defer image.Close()
+	return image.MirrorDemote()
+}
+
+func (rv *rbdVolume) resyncImage(ctx context.Context) error {
+	util.DebugLog(ctx, "rbd: resync image %s", rv.String())
+	image, err := rv.open()
+	if err != nil {
+		return err
+	}
+	defer image.Close()
+	return image.MirrorResync()
+}
+
+type imageMirrorStatus struct {
+	Name        string `json:"name"`
+	State       string `json:"state"`
+	Description string `json:"description"`
+	LastUpdate  string `json:"last_update"`
+}
+
+// Fix me once https://github.com/ceph/go-ceph/issues/460 is fixed.
```

Comment:
please use standard `FIXME:` tagging so that https://www.tickgit.com/browse?repo=github.com/ceph/ceph-csi can be used to find it